### PR TITLE
Automatic detection of dec (. or ,)

### LINKED ISSUE #2431: Automatic detection of dec=',' in Europe
I'm not sure what `base` and `readr` do in this regard, but currently in `fread`, `dec=='.'` by default and needs manually setting to `','` in Europe for numerics with comma as the decimal separator. It could instead be detected automatically like `sep` already is. Please +1 this issue if you'd like this.

Further, `dec` could be automatically detected per-column for files where some numeric columns use `','` and other columns use `'.'`.  But does anyone need that?

### PR DESCRIPTION
**Pull request by MichaelChirico**
Closes #2431

The idea is to shoehorn `dec` detection into column type detection -- whenever we reach testing for double (or double extended), we take two passes. First with `dec='.'`, then with `dec=','`, we count a "win" for `dec=x` if the field parsed succesfully with `x`. At the end of all the jumps, we assign `dec` based on the "winner".

The example provided by Jan is passing:

```
fread("a;b\n1,5;2,5")
#        a     b
#    <num> <num>
# 1:   1.5   2.5
```

TODO:

 - [x] Add new tests
 - [x] Documentation, NEWS item

### CODE REVIEW

File: inst/tests/tests.Rraw [hunk 1]
```diff
@@ -2544,15 +2544,13 @@ if (test_bit64) {
...
-  DT[,a2:=as.integer64(a)][,a3:=as.double(a)][,a4:=gsub(" ","",format(a))]
-  DT[,b2:=as.double(b)][,b3:=gsub(" ","",format(b))]
-  DT[,r:=a/100][,r2:=gsub(" ","",format(r))]
-  DT[112, a2:=as.integer64(12345678901234)]   # start on row 112 to avoid the first 100
-  DT[113, a3:=3.14]
-  DT[114, a4:="123A"]
-  DT[115, b2:=1234567890123.45]
-  DT[116, b3:="12345678901234567890A"]  # A is needed otherwise read as double with loss of precision (TO DO: should detect and bump to STR)
-  DT[117, r2:="3.14A"]
+  DT[ , a2 := as.integer64(a)][112L, a2 := as.integer64(12345678901234)]   # start on row 112 to avoid the first 100
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:2547]: no change to these tests, just syntax -- hard to read/understand what's going on as is

**Comment by MichaelChirico** [comment]
This entire section of `?fread` looks completely outdated? I don't think any of it applies to current master:

```
\bold{Decimal separator and locale:} \code{fread(\dots,dec=",")} should just work. \code{fread} uses C function \code{strtod} to read numeric data; e.g., \code{1.23} or \code{1,23}. \code{strtod} retrieves the decimal separator (\code{.} or \code{,} usually) from the locale of the R session rather than as an argument passed to the \code{strtod} function. So for \code{fread(\dots,dec=",")} to work, \code{fread} changes this (and only this) R session's locale temporarily to a locale which provides the desired decimal separator.

On Windows, "French_France.1252" is tried which should be available as standard (any locale with comma decimal separator would suffice) and on unix "fr_FR.utf8" (you may need to install this locale on unix). \code{fread()} is very careful to set the locale back again afterwards, even if the function fails with an error. The choice of locale is determined by \code{options()$datatable.fread.dec.locale}. This may be a \emph{vector} of locale names and if so they will be tried in turn until the desired \code{dec} is obtained; thus allowing more than two different decimal separators to be selected. This is a new feature in v1.9.6 and is experimental. In case of problems, turn it off with \code{options(datatable.fread.dec.experiment=FALSE)}.
```

Shall I just wipe it altogether?

File: man/fread.Rd [hunk 2]
```diff
@@ -77,9 +77,9 @@ If an empty line is encountered then reading stops there with warning if any tex
...
 
-\bold{Decimal separator and locale:} \code{fread(\dots,dec=",")} should just work. \code{fread} uses C function \code{strtod} to read numeric data; e.g., \code{1.23} or \code{1,23}. \code{strtod} retrieves the decimal separator (\code{.} or \code{,} usually) from the locale of the R session rather than as an argument passed to the \code{strtod} function. So for \code{fread(\dots,dec=",")} to work, \code{fread} changes this (and only this) R session's locale temporarily to a locale which provides the desired decimal separator.
+\bold{Decimal separator:} \code{dec} is used to parse numeric fields as the separator between integral and fractional parts. When \code{dec='auto'}, during column type detection, when a field is a candidate for being numeric (i.e., parsing as lower types has already failed), \code{dec='.'} is tried, and, if it fails to create a numeric field, \code{dec=','} is tried. At the end of the sample lines, if more were successfully parsed with \code{dec=','}, \code{dec} is set to \code{','}; otherwise, \code{dec} is set to \code{'.'}.
```
Comment (MichaelChirico) [inline man/fread.Rd:80]: These two paragraphs were totally irrelevant to current `fread` behavior, AFAICT.

**Comment by MichaelChirico** [comment]
Works well on all the sample files I've seen so far -- from Vilmantas in #2431 and from [Tim A. on Twitter](https://twitter.com/TimSalabim3/status/1264890054687764485) (I used table ID `32211-02-02-4` from [here](https://www.regionalstatistik.de/genesis/online/data?operation=statistic&levelindex=0&levelid=1590419685414&code=32271)

**Comment by jangorecki** [comment]
Feels good, any benchmark yo measure overhead of new feature?

**Comment by MichaelChirico** [comment]
Sure, how about:

Screenshot from 2024-04-04 12-50-44

Not quite sure how to interpret every bar / how wide confidence bands are here. But it seems clear (and intuitive) that when `dec=','` is detected there's massive upside from being able to parallelize reading numerics instead of strings.

Looks like basically a wash besides that, roughly as I'd expect (this PR only affects the column sampling phase, which should be a small sliver of overall read on large files).

<details>

```r
library(data.table)
library(microbenchmark)

params = CJ(nrow = 10^c(2,4,6), ncol=10^(1:2), pct_dot_sep=c(0, .2, .8, 1))
params[, nrep := pmin(1e9 / nrow / ncol, 1000)]

get_median_runtime <- function(nrow, ncol, pct_dot_sep, nrep) {
  tmp <- tempfile()
  on.exit(unlink(tmp))

  outDT = setDT(replicate(ncol, rnorm(nrow), simplify=FALSE))
  dec_cols = sample(ncol(outDT), pct_dot_sep * ncol(outDT))
  for (jj in dec_cols) set(outDT, , jj, formatC(outDT[[jj]], decimal.mark=","))
  fwrite(outDT, tmp, sep=";")
  median(replicate(nrep, {
    t0=get_nanotime()
    fread(tmp, sep=";")
    get_nanotime()-t0
  }))
}

system('git checkout master')
pkgload::load_all()
params[, by = .I, median_runtime_master := get_median_runtime(nrow, ncol, pct_dot_sep, nrep)]

system('git checkout fread-auto-dec')
pkgload::load_all()
params[, by = .I, median_runtime_branch := get_median_runtime(nrow, ncol, pct_dot_sep, nrep)]

# NB: this could be cleaned up a bit...
params = melt(params, idcol = c("nrow", "ncol", "pct_dot_sep"), measure.vars = patterns("^median_runtime"))
setnames(params, c("variable", "value"), c("branch", "median_runtime"))
params[, branch := gsub("^median_runtime_", "", branch)]
params[branch=="branch", branch := "fread-auto-dec"]
params=dcast(params, nrow+ncol+pct_dot_sep~branch,value.var='median_runtime')
params[, improvement:=`fread-auto-dec`/master-1]

params[,
  dcast(.SD, pct_dot_sep ~ nrow+ncol, value.var='improvement')
][,
  .SDcols=!'pct_dot_sep',
  barplot(as.matrix(.SD), beside=TRUE, main='This PR / master\nnrow_ncol\npct_dot_sep=0, 1/5, 4/5, 1')
]
```

</details>

**Comment by MichaelChirico** [comment]
Hi @jangorecki, anything left to do here?

**Review by jangorecki — APPROVED** [review]

**Pull request merged**